pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,383 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/meta/metatypes"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const ResourceResyncTime = 30 * time.Second
+
+type Monitor struct {
+	Store      cache.Store
+	Controller *framework.Controller
+}
+
+type SelfReference struct {
+	metatypes.OwnerReference
+	// This is needed by the dynamic client
+	Namespace string
+}
+
+type node struct {
+	identity   SelfReference
+	dependents []*node
+	// When processing an Update event, we need to compare the updated
+	// ownerReferences with the owners recorded in the graph.
+	owners []metatypes.OwnerReference
+}
+
+type EventType string
+
+const Add EventType = "Add"
+const Update EventType = "Update"
+const Delete EventType = "Delete"
+
+type Event struct {
+	Type   EventType
+	Obj    interface{}
+	OldObj interface{}
+}
+
+type Propagator struct {
+	eventQueue  *workqueue.Type
+	graphLookup map[types.UID]*node
+	gc          *GarbageCollector
+}
+
+func (p *Propagator) addToGraphLookup(n *node) {
+	p.graphLookup[n.identity.UID] = n
+}
+
+func (p *Propagator) removeFromGraphLookup(n *node) {
```

Comment:
removeNode

If this map doesn't need to be locked, please state why in a comment in the struct.